TST: _lib: improve array API assertions

mdhaber opened the pull request:
Suggestions for scipy/scipy/#19186

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -196,19 +196,26 @@ def is_torch(xp):
+                  check_namespace=True, check_dtype=True, check_shape=True):
+    if check_namespace:
         _assert_matching_namespace(actual, desired)
-    xp = array_namespace(actual)
```
might as well pass this in

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -196,19 +196,26 @@ def is_torch(xp):
     if check_dtype:
         assert_(actual.dtype == desired.dtype,
-                f"Desired dtype: {desired.dtype}, actual dtype: {actual.dtype}")
+                "dtypes do not match.\n"
```
The particular format doesn't matter, but we might as well pick a *consistent* format for all of these messages.

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -196,19 +196,26 @@ def is_torch(xp):
+                "Shapes do not match.\n"
+                f"Actual: {actual.shape}\n"
+                f"Desired: {desired.shape}")
+        _check_scalar(actual, desired, xp)
```
I originally suggested that `_check_scalar` would be part of `check_namespace`/`check_xp`, but I think it's more appropriate for it to be part of the shape check (or something separate altogether).

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -217,26 +224,43 @@ def _assert_matching_namespace(actual, desired):
+    # Therefore, regardless of whether the developer provides a scalar or 0D
+    # array for `desired`, we would typically want the type of `actual` to be
+    # the type of `desired[()]`. If the developer wants to override this
+    # behavior, they can set `check_shape=False`.
```
Does this make sense?

lucascolley replied:
Yes I think so, although modulo CuPy https://github.com/scipy/scipy/pull/19186#discussion_r1320502626?

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -217,26 +224,43 @@ def _assert_matching_namespace(actual, desired):
-    _strict_check(actual, desired, check_xp=check_xp,
-                  check_dtype=check_dtype, check_shape=check_shape)
     if xp is None:
         xp = array_namespace(actual)
```
Since `xp` is going to be determined anyway, might as well share it with `_strict_check`.

mdhaber reviewed (commented):
Some self-review and questions.

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -217,26 +224,43 @@ def _assert_matching_namespace(actual, desired):
+
+
+def _check_scalar(actual, desired, xp):
+    # Shape check alone is sufficient unless desired.shape == (). Also,
+    # only NumPy and CuPy distinguish between scalars and arrays, AFAICT.
+    if desired.shape != () or not (is_numpy(xp) or is_cupy(xp)):
```
```suggestion
    # Shape check alone is sufficient unless desired.shape == (). Also,
    # only NumPy distinguishes between scalars and arrays.
    if desired.shape != () or not is_numpy(xp):
```

mdhaber commented on scipy/_lib/tests/test_array_api.py:
```diff
@@ -64,3 +66,48 @@ def test_copy(self, xp):
+            with pytest.raises(AssertionError, match="Types do not match."):
+                xp_assert_equal(xp.asarray(0.), xp.float64(0))
+            xp_assert_equal(xp.float64(0), xp.asarray(0.))
+        else:
+            with pytest.raises(AssertionError, match="Types do not match."):
+                xp_assert_equal(xp.float64(0), xp.asarray(0.))
+            xp_assert_equal(xp.asarray(0.), xp.float64(0))
```
```suggestion
```

mdhaber commented on scipy/_lib/tests/test_array_api.py:
```diff
@@ -64,3 +66,48 @@ def test_copy(self, xp):
+
+    @array_api_compatible
+    def test_check_scalar(self, xp):
+        if not (is_numpy(xp) or is_cupy(xp)):
```
```suggestion
        if not is_numpy(xp):
```

mdhaber reviewed (commented):
OK, I guess these things can be removed since CuPy doesn't really distinguish. `cupy.float32`, e.g., exists, but it's just an alias for `np.float32`.

lucascolley reviewed (changes requested):
Hi Matt, this looks great overall! However, it is giving quite a few test fails.

Here is a patch which fixes most of them:

```diff
diff --git a/scipy/_lib/_array_api.py b/scipy/_lib/_array_api.py
index 82ab793d8..ca5e8c471 100644
--- a/scipy/_lib/_array_api.py
+++ b/scipy/_lib/_array_api.py
@@ -201,7 +201,6 @@ def _strict_check(actual, desired, xp,
     if check_namespace:
         _assert_matching_namespace(actual, desired)
 
-    actual = xp.asarray(actual)
     desired = xp.asarray(desired)
 
     if check_dtype:
@@ -217,6 +216,8 @@ def _strict_check(actual, desired, xp,
                 f"Desired: {desired.shape}")
         _check_scalar(actual, desired, xp)
 
+    return actual, desired
+
 
 def _assert_matching_namespace(actual, desired):
     actual = actual if isinstance(actual, tuple) else (actual,)
@@ -259,8 +260,9 @@ def xp_assert_equal(actual, desired, check_namespace=True, check_dtype=True,
                     check_shape=True, err_msg='', xp=None):
     if xp is None:
         xp = array_namespace(actual)
-    _strict_check(actual, desired, xp, check_namespace=check_namespace,
-                  check_dtype=check_dtype, check_shape=check_shape)
+    actual, desired = _strict_check(actual, desired, xp,
+                                    check_namespace=check_namespace,
+                                    check_dtype=check_dtype, check_shape=check_shape)
     if is_cupy(xp):
         return xp.testing.assert_array_equal(actual, desired, err_msg=err_msg)
     elif is_torch(xp):
```

There are two failures remaining though:

<details>

```
___________________________________________________________________________ TestArrayAPI.test_strict_checks[shape0-int64-torch] ____________________________________________________________________________
scipy/_lib/tests/test_array_api.py:91: in test_strict_checks
    with pytest.raises(AssertionError, match="dtypes do not match."):
E   Failed: DID NOT RAISE <class 'AssertionError'>
        dtype      = torch.int64
        options    = {'check_dtype': True, 'check_namespace': False, 'check_shape': False}
        self       = <scipy._lib.tests.test_array_api.TestArrayAPI object at 0x7f7b05c45690>
        shape      = ()
        x          = tensor(1)
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
        y          = 1
___________________________________________________________________________ TestArrayAPI.test_strict_checks[shape1-int64-torch] ____________________________________________________________________________
scipy/_lib/tests/test_array_api.py:92: in test_strict_checks
    xp_assert_equal(x, y, **options)
        dtype      = torch.int64
        options    = {'check_dtype': True, 'check_namespace': False, 'check_shape': False}
        self       = <scipy._lib.tests.test_array_api.TestArrayAPI object at 0x7f7afc86bd50>
        shape      = (3,)
        x          = tensor([1, 1, 1])
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
        y          = 1
scipy/_lib/_array_api.py:272: in xp_assert_equal
    return xp.testing.assert_close(actual, desired, rtol=0, atol=0, equal_nan=True,
E   AssertionError
        actual     = tensor([1, 1, 1])
        check_dtype = True
        check_namespace = False
        check_shape = False
        desired    = tensor(1)
        err_msg    = ''
        xp         = <module 'scipy._lib.array_api_compat.array_api_compat.torch' from '/home/lucas/dev/myscipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/array_api_compat/torch/__init__.py'>

During handling of the above exception, another exception occurred:
scipy/_lib/tests/test_array_api.py:91: in test_strict_checks
    with pytest.raises(AssertionError, match="dtypes do not match."):
E   AssertionError: Regex pattern did not match.
E    Regex: 'dtypes do not match.'
E    Input: ''
        dtype      = torch.int64
        options    = {'check_dtype': True, 'check_namespace': False, 'check_shape': False}
        self       = <scipy._lib.tests.test_array_api.TestArrayAPI object at 0x7f7afc86bd50>
        shape      = (3,)
        x          = tensor([1, 1, 1])
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
        y          = 1
========================================================================================= short test summary info ==========================================================================================
FAILED scipy/_lib/tests/test_array_api.py::TestArrayAPI::test_strict_checks[shape0-int64-torch] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED scipy/_lib/tests/test_array_api.py::TestArrayAPI::test_strict_checks[shape1-int64-torch] - AssertionError: Regex pattern did not match.
======================================================================================= 2 failed, 30 passed in 1.38s =======================================================================================
```

</details>

The first one is because `torch.int64 != dtype('int64')`. I'm not sure about the second one.

Can you try to fix these last two (and apply my patch if it still applies)? Then this will LGTM!

mdhaber commented:
I had added `actual = xp.asarray(actual)` at the last minute and didn't test afterwards. Oops! Not sure what I was thinking.

Why return `actual` and `desired`?

lucascolley commented:
> I had added actual = xp.asarray(actual) at the last minute

That's my fault, I added it previously to try to deal with the case of actual being a scalar, before you made the changes for scalars which mostly fixed things.

> Why return actual and desired?

We can probably just `return desired`. We need this because we coerce with `desired = xp.asarray(desired)`. Without it, we can end up trying to test between different namespaces in the rest of `_assert_matching_namespace`, which fails.

> I didn't catch the torch errors because I'm having trouble testing those. Did you say that the latest release of PyTorch fixes something?

Not yet released, there is a fix for arrays not being created on GPU when they should be. That isn't relevant here though; what problems are you having?

mdhaber commented:
> That's my fault, I added it previously

Not really, because I had removed it since it wasn't needed : )

> Without it, we can end up trying to test between different namespaces

Ok. The reason I ask is that this is essentially what I was pushing for in the original PR - automatic conversion to the appropriate backend. 

> there is a fix for arrays not being created on GPU when they should be.

Almost all of my torch tests are failing, masking the two real failures. I'll send the message when I'm at a computer.

lucascolley commented:
> Maybe now with check_namespace=True by default, it resolves concerns with adding the automatic conversion. Now, the user has to do the conversion themselves to pass the namespace check. However, they can disable the check, and only then does the conversion do anything for them.

Yeah, this seems much better - I'm certainly happy with that.

mdhaber commented:
> Almost all of my torch tests are failing,

Never mind. They are probably failing for the reason you pointed out. Originally, I assumed they were due to the bug you said they are going to fix and ignored it, since it seems strange to me that `torch.testing` can't compare a tensor against a `np.int32`.

mdhaber commented:
I think the second one is because `torch` doesn't like comparing arrays of different sizes. 
```
torch.testing.assert_close(torch.asarray([1, 1, 1]), torch.asarray(1))
# AssertionError: The values for attribute 'shape' do not match: torch.Size([3]) != torch.Size([]).
```
So after performing the shape check, I broadcasted the shape of `desired` to the shape of `actual`. 

I don't think this was new, right? The test just caught a lurking bug?

lucascolley reviewed (approved):
This looks good to me now, just this typo below. I'll merge after that and we can finalize on the main repo.

Can confirm that `python dev.py test -t scipy/_lib/tests/test_array_api.py -b all` passes.

mdhaber commented on scipy/_lib/tests/test_array_api.py:
```diff
@@ -64,3 +66,44 @@ def test_copy(self, xp):
+    @array_api_compatible
+    def test_check_scalar(self, xp):
+        if not is_numpy(xp):
+            pytest.skip("Scalars only exist in NumPy in CuPy")
```
Well I just forgot to remove "and CuPy". CuPy does have scalars, but they happen to be duplicates of NumPy's.

mdhaber commented:
It just needs a `[()]` either at the end of the function or in the test. I left it out of the function on purpose because there are some places in SciPy that do return 0d arrays. I don't know if that will cause any tests to fail, but it should be simple to make everything happy.

lucascolley commented:
Thanks, looks like https://github.com/scipy/scipy/commit/0c3849f51db40987a7cfffc9cb661ca0325b7fb4 has done it.
